### ISSUE TITLE
Uses global config but can work with only system config

### ISSUE THREAD

**Issue by NickVeld** [desktop/desktop#12145]
### Describe the bug

> A clear and concise description of what the bug is.

When you change your global user settings of `git` Github starts using them and it is annoying instead of settings in the options of the app.

Each user can have several Git-based VCS but GitHub Desktop with its bundled `git` relys on the global config and prevents `Git For Windows` or other `git` apps from usage this config.

### Version & OS

> Open 'About GitHub Desktop' menu to see the Desktop version. Also include what operating system you are using.

* 2.8.1
* Win10

### Steps to reproduce the behavior

> 1. Go to '...'
> 2. Click on '....'
> 3. Scroll down to '....'
> 4. See error

1. Change the email to different from attached to the current Github account in the "user" section in the global config (`~/.gitconfig`, where `~` is users' home directory).

### Expected behavior

> A clear and concise description of what you expected to happen.

GitHub Desktop uses the `--author` flag of `git` (e.g `git commit --author="Username <Username@users.noreply.github.com>" -m "Message"`) or only the "system-wide" config (stored in `~\AppData\Local\GitHubDesktop\app-x.x.x\resources\app\git\etc\gitconfig`) (and the repository-wide config if it exists)

### Actual behavior

> A clear and concise description of what actually happened.

1. After the change of the email in the global config see this icon instead of the avatar of the profile
![image](https://user-images.githubusercontent.com/17028401/117150199-99c3bc80-adc0-11eb-9b2f-2983d8e34fbf.png)
2. Click on this icon
3. See this window
![image](https://user-images.githubusercontent.com/17028401/117154070-32a80700-adc4-11eb-8f5b-09bf2b885c7d.png)

From this I assume that GitHub Desktop uses the default global gitconfig values but not the `--author` flag of `git`.

### Screenshots

> Add screenshots to help explain your problem, if applicable.

Attached above.

### Logs

> Attach your logs by opening the `Help` menu and selecting `Show Logs...`, if applicable.

Not applicable

### Additional context

> Add any other context about the problem here.

Since Github Desktop has own bundled git, it can store all configuration in the own "system-wide" config (stored in `~\AppData\Local\GitHubDesktop\app-x.x.x\resources\app\git\etc\gitconfig`). I tried to move all lines from the global config to the "system-wide" one. And it works as long as no set "[user]" section in the global config.  However it is not really good option for the multi-user system.

So it is better to force bundled `git` search the global config in the path different of the orginal global config, e.g. `~/GitHubDesktop/.gitconfig' or '~/.githubdesktopconfig`. The easiest solution to use the values from the GitHubDesktop options (Git>Name,Email) overwritable  by the repository-wide config in the `--author` flag (e.g `git commit --author="Username <Username@users.noreply.github.com>" -m "Message"`)

The continuation of https://github.com/desktop/desktop/issues/5017

**Comment by NickVeld** [comment]
### Workaround for the systems that have multiple `git` apps

**Warning**: this workaround is applicable only if you do not care about unintentional(?) usage and awareness of your name and email for `git` by other users of the system.

**Warning**: after each update `app-x.x.x` changes thus while Github Desktop is updating the config maybe need changing in each new `app-x.x.x`

You can move needed lines from the global config (`~/.gitconfig`) to the "system-wide" one (for Github Desktop `~\AppData\Local\GitHubDesktop\app-x.x.x\resources\app\git\etc\gitconfig`, for other `GIT_PATH\etc\gitconfig`) customizing them according to your needs. In this way different for each `git` app user's name and email can be used.

**Comment by sergiou87** [comment]
Hi @NickVeld! 👋  Thank you for your feedback ❤️ 

> GitHub Desktop with its bundled `git` relys on the global config and prevents `Git For Windows` or other `git` apps from usage this config.

Why do you mean by this exactly? The way GitHub Desktop uses `.gitconfig` shouldn't affect `Git For Windows`. Desktop doesn't store a user name or email, that info is always read from / written to the `.gitconfig` file. If you change `user.name` or `user.email` from the command line, it should affect Desktop too (if you use the `--global` parameter).

As you guessed, GitHub Desktop uses whatever `.gitconfig` file is more relevant to your repository. Most of the times, this will be the global config from `~/.gitconfig`, but it can also be the local `.gitconfig` in your repo. You can set a specific name/email for your repository from the `Repository` -> `Repository Settings…` -> `Git config` screen:

![image](https://user-images.githubusercontent.com/1083228/117458358-82183f80-af4a-11eb-8f5a-8ff6a9502b0f.png)

Another advanced option is to use conditional includes to specify different name / email depending on the folder in which the project is located. Take a look at #12041 for more info (although as you can see, that scenario is still not perfectly supported 😅 ).

Do any of these suggestions help you with your issue?

If not, could you explain more in detail the scenario and the problem you're running into?

**Comment by NickVeld** [comment]
Hi

> Desktop doesn't store a user name or email, that info is always read from / written to the .gitconfig file

![image](https://user-images.githubusercontent.com/17028401/117472150-43da4a80-af61-11eb-809d-f9e3f3780905.png)

So how Github Desktop show me this warn instantly and without the Internet connection if there is no name/email is stored by it?
But if the user id can be checked (and consequently it is stored somewhere) then 

> it can also be the local .gitconfig in your repo

I know about it and mentioned it in my initial post using the expression "the repository-wide config".

I or someone else could manage many repos and reconfigurating each of them is an option but not really nice.

> Do any of these suggestions help you with your issue?

I did not intend to discuss my struggles with the issue and as you can see I resolved it myself with the workaround written above.
I prepared for developers the extensive report and listed my suggestions in it that can make GitHub Desktop more user-friendly.
It is just up to the developer team to implement it or not.

Just so we are clear, I did not want any support about this issue initially, I wanted only the enhancement described by me in the product.

**Comment by sergiou87** [comment]
The info of the signed in users (on GitHub and GitHub Enterprise) is indeed persisted by the app, but not what's used for the commits. Sorry if I wasn't clear 😓 

> I did not intend to discuss my struggles with the issue

We're just trying to understand what the underlying problem is, see how it could relate to other users and try to find a solution to fix that problem. We appreciate your suggestion, but it's hard to evaluate it if we don't understand this whole scenario 😅 

So basically we'd love to understand…
- What were you trying to achieve?
- Why did you need to change your Git config?
- When you got the warning, why was it incorrect?

Thank you again for your time!

**Comment by NickVeld** [comment]
> What were you trying to achieve?

The same that mentioned not only in the my issue but in the attached other issue #5017 :

Use different account settings (name, email) on Github Desktop from other `git` apps

> Why did you need to change your Git config?

Because I need another different `git` account separate from the pair(s) of name+email registered in the GitHub account

> When you got the warning, why was it incorrect?

The warning is correct for the current implementation. But the current implementation prevents me from achieving the written above because GitHub Desktop have all possibilities easily to work without a global configuration file but use it. (Changing each repository settings to use `.gitconfig` of the repository is a solution but it is a not good user experience.)

**Comment by sergiou87** [comment]
Thank you for that extra context, it's very useful!

> Use different account settings (name, email) on Github Desktop from other `git` apps

Let me be more specific: if you're working on a repository that is hosted on GitHub, where you have a user called `NickVeld` which has a bunch of email accounts registered… why would you want to use a non-registered email account to make commits in that repository?

If you did so, the commits in that repository would appear [wrongly attributed on GitHub.com](https://docs.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) (hence the warning you're seeing).

And if that wasn't a GitHub repository, you shouldn't see the warning at all, since the concept of "commit attribution" doesn't even apply.

**Comment by NickVeld** [comment]
> why would you want to use a non-registered email account to make commits in that repository

Sometimes I want to push into non-GitHub repositories that demand the attribution not registered in GitHub.

> And if that wasn't a GitHub repository, you shouldn't see the warning at all

But I do not want to either change the global settings each time when I switch between GitHub and non-GitHub repositories nor set up each repository/ Especially, when I know that there is easy solution for GitHub Desktop to save the developer's (including me) time just not touching global `git` config and using internal knowledge about accounts or using own configuration file.

**Comment by sergiou87** [comment]
We took that scenario into consideration, and is the reason why we added the ability to change it locally in every repository.

I understand how that solution is not convenient to you, but we still think it's the best solution for _most_ of GitHub Desktop users out there. Using a specific `.gitconfig` file just for GitHub Desktop would have different drawbacks in other circumstances for people relying on the current behavior. In the end there are no perfect solutions, everything is a trade off 😅 

With that in mind, I think the best and most compatible workaround is using conditional includes as in #12041, specially after we fix the existing issue.

Thank you again for your time reporting this issue and providing context to us about it!! ❤️

**Comment by NickVeld** [comment]
@sergiou87 

> Using a specific .gitconfig file just for GitHub Desktop would have different drawbacks in other circumstances for people relying on the current behavior.

Could you, please, provide the example of the "relying on the current behavior"? Is somebody parsing it automatically (if it is so then what is its application?) or what?

I am sure that if you care  about the backward compatibility, you can just add a simple switch between the behaviors easily and set the old behavior as the default one, gather the statistics and make an informed decision about the subject .

**Comment by sergiou87** [comment]
> Could you, please, provide the example of the "relying on the current behavior"? Is somebody parsing it automatically (if it is so then what is its application?) or what?

A simple example of this is people who normally switch between GitHub Desktop and command line git and expect both of them to use the same user name and email.

A more convoluted example is where more advanced users who nowadays leverage this use of the global `.gitconfig` file to set up commit signing: https://gist.github.com/xavierfoucrier/c156027fcc6ae23bcee1204199f177da

> you can just add a simple switch between the behaviors easily and set the old behavior as the default one

That would be an option, but we'd rather not add more settings unless we think they're needed, and try to keep the app as simple as possible 😳

**Comment by NickVeld** [comment]
> A simple example of this is people who normally switch between GitHub Desktop and command line git and expect both of them to use the same user name and email.

Do you always expect that your account used in the app A automatically used in the app B because the app B just can support signing in with this account?
I am personally hate when any app signs me in without my permission. Because it causes not intentional history records and synchronizations.

Moreover, why two absolutely differently application (Git for Windows and bundled `git` from GitHub Desktop) from two different development teams should be set up in one place? It is absolutely not convenient and used sometimes only as a temporary solution in quite small specific applications.

No one asks to break current setups immediately, there are plenty options to implement the solution of this issue with the backward compatability.
One more solution: insert one more item into current hierarchy (repo config > global config > system config) and get repo config > github config > global config > system config

> this use of the global .gitconfig file to set up commit signing

It is not an example, it is instruction for the current environment. When a separate config appears, this instruction will stay working and also applicable for the new config.

> and try to keep the app as simple as possible

I am sure that the solution of this issue is not more complex than most things in Github Desktop
